### PR TITLE
Fix external pagemap usage.

### PR DESCRIPTION
**Pull request by davidchisnall**
At some point in the refactoring, these were broken.

### CODE REVIEW

**Comment by mjp41** [comment]
Could you add unit tests, so these don't break in the future?

**Comment by davidchisnall** [comment]
I've been pondering how to do this.  These only show up when snmalloc is used as the malloc implementation in libc and also in another library.

**Comment by mjp41** [comment]
So we have tests that build two different allocators, could you modify that to cover this case:
https://github.com/microsoft/snmalloc/tree/master/src/test/func/two_alloc_types

Does it have to be in libc?

**Comment by davidchisnall** [comment]
Possibly not.  I'll have a look at whether we can make one of those find the other's pagemap.

**Comment by davidchisnall** [comment]
I've added a unit test.  This is blocking work in Verona, so it would be good to get it merged soon.

File: src/test/func/external_pagemap/external_pagemap.cc [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+	auto &p = ExternalGlobalPagemap::pagemap();
+	auto &global = GlobalPagemap::pagemap();
+	assert(&p == &global);
+	// Get a valid heap address
+	uintptr_t addr = reinterpret_cast<uintptr_t>(malloc(42));
+	// Make this very strongly aligned
+	addr &= ~0xfffffULL;
+	void *page = p.page_for_address(addr);
+	assert(page == p.page_for_address(addr + 128));
+	assert(page == p.page_for_address(addr + 128));
```
Comment (mjp41) [inline src/test/func/external_pagemap/external_pagemap.cc:15]: Duplicate assert?

File: src/test/func/external_pagemap/external_pagemap.cc [hunk 2]
```diff
@@ -0,0 +1,21 @@
+#define SNMALLOC_EXPOSE_PAGEMAP 1
+#  include <override/malloc.cc>
+
+int main()
+{
+	auto &p = ExternalGlobalPagemap::pagemap();
+	auto &global = GlobalPagemap::pagemap();
+	assert(&p == &global);
```
Comment (mjp41) [inline src/test/func/external_pagemap/external_pagemap.cc:8]: Can you use `SMALLOC_CHECK` in the test please.  This should be checked in release builds.

**Review by mjp41 — APPROVED** [review]
Looks good minor fixes

File: src/test/func/external_pagemap/external_pagemap.cc (hunk 2 above)
Comment (davidchisnall) [reply src/test/func/external_pagemap/external_pagemap.cc:8]: `SNMALLOC_CHECK` was added after this PR was originally raised.  I've rebased it on master.

File: src/test/func/external_pagemap/external_pagemap.cc [hunk 3]
```diff
@@ -0,0 +1,30 @@
+#if defined(SNMALLOC_PASS_THROUGH) || defined(_WIN32)
```
Comment (mjp41) [inline src/test/func/external_pagemap/external_pagemap.cc:1]: If you set a `SNMALLOC_NAME_MANGLE` then you can make the test work on Windows.

Comment (davidchisnall) [reply src/test/func/external_pagemap/external_pagemap.cc:1]: That's not the problem.  The Windows declarations for the standard malloc functions have a bunch of declspecs in them that we're not using and this causes compile failures.

Comment (mjp41) [reply src/test/func/external_pagemap/external_pagemap.cc:1]: If you had set a name mangle then they wouldn't conflict.

**Pull request merged**